### ISSUE TITLE
Support the Preserve Case Formatter

### ISSUE THREAD

**Issue by guydc** [solo-io/gloo#5522]
**Is your feature request related to a problem? Please describe.**
Envoy [normalizes](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/header_casing#http-1-1-header-casing) headers to lower-case. In some situations, the original casing should be preserved (lower, proper or mixed). Gloo Edge currently supports the [proper case formatter](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options/hcm/hcm.proto.sk/), but not the preserve case formatter. 

**Describe the solution you'd like**
Expose the preserve case formatter in HCM options.

**Comment by asayah** [comment]
we can use stateful https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-msg-config-core-v3-http1protocoloptions-headerkeyformat

**Comment by sam-heilbron** [comment]
PR that exposed ProperCaseFormatter: https://github.com/solo-io/gloo/pull/2986

As Adam points out above, we already expose `proper_case_words` and now we need to also expose `stateful formatter`.

**Comment by guydc** [comment]
dupe: https://github.com/solo-io/gloo/issues/3274

**Comment by guydc** [comment]
Hi @gunnar-solo, @sam-heilbron. I tried out this feature, and header casing is still not preserved. I tested it locally with envoy, and it appears that preservation of response header casing will only occur if the formatter is configured in both HCM and Cluster context. Is there a way to apply this config to a Gloo upstream?

**Comment by gunnar-solo** [comment]
Fix is in/merged to `master` and `v1.10.x`.  It's also been released with `v1.10.12`

Manual test/verification steps:
```bash
# install `gloo@v1.10.12` to a fresh test cluster
kind create cluster
helm install -n gloo-system gloo gloo/gloo --version 1.10.12 --create-namespace

# apply a custom-made response header echoing service and configure it with `gloo`
kubectl apply -f https://raw.githubusercontent.com/gunnar-solo/python-header-echo/master/manifests.yaml
kubectl apply -f - <<EOF
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: header-echo-service
  namespace: gloo-system
spec:
  virtualHost:
    domains:
      - 'header-echo-service'
    routes:
      - matchers:
          - prefix: /
        routeAction:
          single:
            upstream:
              name: default-header-echo-service-8000
              namespace: gloo-system
EOF

# set feature flags on misc. gloo objects
kubectl -n gloo-system get Gateway gateway-proxy -ojson | jq '.spec.httpGateway.options.httpConnectionManagerSettings["preserveCaseHeaderKeyFormat"]=true' | kubectl apply -f -
kubectl -n gloo-system get Upstream default-header-echo-service-8000 -ojson | jq '.spec.connectionConfig.http1ProtocolOptions["preserveCaseHeaderKeyFormat"]=true' | kubectl apply -f -

# expose gateway-proxy and verify header case-preservation
kubectl port-forward -n gloo-system deploy/gateway-proxy 8080
curl -v -H "Host: header-echo-service" -H "aA-Aa-AA-aa: grep_me" localhost:8080/headers | jq
```
